### DISCUSSION TITLE
Several types of BC on the same border

### DISCUSSION THREAD

**Discussion by DobryVecher1** [zwicker-group/py-pde#365] [Q&A] [answered]
Hello, thank you for the great package! 

I have the following question: for example, I have a rectangular mesh and I need to impose BC of different types for certain parts of the upper wall (see figure). Is it possible to do with py-pde?
<img width="500" alt="Screenshot_2" src="https://user-images.githubusercontent.com/47660072/220083615-383a2fac-c749-4048-a915-a47d436040c3.png">

**Comment by david-zwicker** [comment, marked as answer]
This should be possible in principle, but is not very accessible at this time. Your best bet might be using an [`ExpressionBC` object](https://py-pde.readthedocs.io/en/latest/packages/pde.grids.boundaries.local.html#pde.grids.boundaries.local.ExpressionBC) for this boundary, where you give a mathematical expression to determine the value of the virtual point. Since this expression can depend on the coordinate along the boundary, it will be in principle possible to implement your idea, but this would involve piecewise functions and is thus not very straight-forward.

In the future, it would be better if we could directly support functions that determine the boundary values. I added an issue for this (#367), but I don't know when we will have time to implement it.

**Reply by david-zwicker** [reply]
I just merged #370, which allows defining functions that specify boundary values or derivatives. Your use case can be covered by using `{"virtual_point": bc_func}`, where `bc_func` is a function calculating the virtual point data:

```python
def bc_func(adjacent_value, dx, x, y, t):
    bc_dirichlet = 2 * BC_VALUE - adjacent_value
    bc_neumann = dx * BC_DERIVATIVE + adjacent_value
    return np.where(x < 0.5, bc_dirichlet, bc_neumann)
```
I hope that helps you getting in the right direction.

**Reply by DobryVecher1** [reply]
Thank you for quick responses, I will try it out
